ee/tabby-webserver/src/repositories.rs

```diff
@@ -0,0 +1,56 @@
+mod resolve;
+
+use anyhow::Result;
+use axum::{extract::Path, http::StatusCode, response::Response, routing, Json, Router};
+use tabby_common::path::repositories_dir;
+use tracing::{instrument, warn};
+
+use crate::{
+    repositories,
+    repositories::resolve::{resolve_dir, resolve_file, resolve_meta, Meta, Repository},
+};
+
+pub fn routers() -> Router {
+    Router::new()
+        .route("/:name/resolve/", routing::get(repositories::resolve))
+        .route("/:name/resolve/*path", routing::get(repositories::resolve))
+        .route("/:name/meta/", routing::get(repositories::meta))
+        .route("/:name/meta/*path", routing::get(repositories::meta))
+}
+
+#[instrument(skip(repo))]
+async fn resolve(Path(repo): Path<Repository>) -> Result<Response, StatusCode> {
+    let root = repositories_dir().join(repo.name_str());
+    let full_path = root.join(repo.path_str());
+    let is_dir = tokio::fs::metadata(full_path.clone())
+        .await
+        .map(|m| m.is_dir())
+        .unwrap_or(false);
+
+    if is_dir {
+        return match resolve_dir(root, full_path).await {
+            Ok(resp) => Ok(resp),
+            Err(err) => {
+                warn!("{}", err);
```

Comment:
Seems major case of error are 404 - not found? Then it's not make sense to write a warning log and returns 500?

Reply:
This Err branch is to handle **io** error which may occur on `read_dir`, `read_dir.next_entry()`.
We call `resolve_dir` only if the path **is** dir and exists, so there's no 404 happen here.

If path is invalid dir, `is_dir` will be `false`, so it's fallback to `resolve_file`.

Reply:
```suggestion
                warn!("failed to resolve_dir <{}>: {}", full_path, err);
```

I see, polishing the error message a bit